cudax/examples/cccl-rt/kernel_launch_patterns.cu

```diff
@@ -0,0 +1,217 @@
+    // one of the threads must construct it before it is used.
+    dynamic_smem_layout& dyn_smem = cudax::dynamic_smem_ref(config);
+
+    // Construct the dynamic_smem_layout object in the shared memory by the first thread in the block.
+    if (config.dims.rank(cudax::thread, cudax::block) == 0)
+    {
+      new (&dyn_smem) dynamic_smem_layout{42};
+    }
+
+    // Wait until the write is finished.
+    __syncthreads();
+
+    // All threads should see the same value in the shared memory.
+    assert(dyn_smem.value == 42);
+  }
+};
+
+#if defined(__CUDACC_EXTENDED_LAMBDA__)
+// Kernel lambda is another form of the kernel functor. It can optionally take the kernel_config as the first argument.
+// Extended lambda are required to use this feature.
+// See https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#extended-lambdas for more info.
+const auto kernel_lambda = [] __device__(auto config, name_buffer name) {
+  say_hello(dim3{config.dims.index(cudax::thread, cudax::block)}, name);
+};
+#endif // defined(__CUDACC_EXTENDED_LAMBDA__)
+
+int main()
+try
+{
+  // Check we have at least one device.
+  if (cuda::devices.size() == 0)
+  {
+    std::fprintf(stderr, "No CUDA devices found\n");
+    return 1;
+  }
+
+  // We will use the first device.
+  cuda::device_ref device = cuda::devices[0];
+
+  // cudax::launch always requires a work submitter, so let's create a CUDA stream.
+  cuda::stream stream{device};
+
+  // Create a custom hierarchy to be used in cudax::launch. We will be launching a 1D grid of 1 block. The block will be
+  // a 2D grid of 2 threads in x and y axis.
+  //
+  // Note that the grid dimensions are passed as template parameters in this example. That means the value can be used
+  // in constexpr context inside the kernel. Block dimensions will be constructed at runtime as usually.
+  const auto hierarchy = cudax::make_hierarchy(cudax::grid_dims<1>(), cudax::block_dims(dim3{2, 2}));
+
+  // Launch an ordinary kernel. cudax::launch takes a stream as the first argument followed by the kernel configuration,
+  // kernel and kernel parameters.
+  cudax::launch(stream, cudax::make_config(hierarchy), kernel, "kernel");
+
+  // Launch a kernel functor. Here, we use cudax::distribute to create the kernel_config for us. This function creates
+  // a simple 1D grid of 1D blocks of a given size.
+  cudax::launch(stream, cudax::distribute<4>(4), kernel_functor{42}, name_buffer{"kernel functor"});
+
+  // Launch a kernel functor that takes a cudax::kernel_config. Note that the kernel config is passed automatically as
+  // the first argument by the cudax::launch function.
+  const auto config =
+    cudax::make_config(hierarchy, cudax::dynamic_shared_memory<kernel_functor_with_config::dynamic_smem_layout>());
+  cudax::launch(stream, config, kernel_functor_with_config{}, name_buffer{"kernel functor with config"});
```

Comment:
suggestion: It's not clear what the purpose of the `cudax::dynamic_shared_memory` config option is. I think it needs more explanation both here and in the kernel itself.